### DISCUSSION TITLE
Feature Request: read_json_auto() on json text field instead of file

### DISCUSSION THREAD

**Discussion by loicalleyne** [duckdb/duckdb#9558] [Ideas]
Why do you want this feature?
I would like to read the JSON of a field stored in a string column as a table without having to COPY to a jsonl file and use read_json_auto(filename).

**Comment by themichaelusa** [comment]
Hard agree, this is immensely useful if you are importing tables w/ mostly scalar and vector columns with some JSON columns that start out as strings or lists of strings. Right now you have to manually define each struct, which for (potentially) heterogenous JSON columns is annoying, and more so that you need to write a ton of boilerplate. 

Boilerplate example here (real code I'm using in production):

 ```
CREATE TEMPORARY MACRO consolidate_metrics() AS TABLE (
    WITH metrics_extract AS (
        SELECT
            id,
            unnest(list_distinct(array_agg(metrics::JSON))) as m
        FROM full_merge_table
        GROUP BY id
    ),
    struct_convert as (
        SELECT id, m::STRUCT(available BOOLEAN, possible BOOLEAN) as st_convert
        FROM metrics_extract
    ),
    jtable as (
        SELECT id, st_convert.* FROM struct_convert
    )
    SELECT 
        id,
        mode(available) as available,
        mode(possible) as possible,
    FROM jtable
    GROUP BY id
);
```

Seems like the bones for this are supported with `json_group_structure` and `read_json_auto` (as @loicalleyne mentioned).

Instead to get a workable table it would look like this:

```
WITH metrics_extract AS (
    SELECT
        id,
        unnest(list_distinct(array_agg(metrics::JSON))) as m
    FROM full_merge_table
    GROUP BY id
)
SELECT id, read_json_auto(m) from metrics_extract
```

**Comment by maitlandmarshall** [comment]
This would be awesome!

Here's how I'm currently converting json to a struct:
duckdb.query("COPY job_details TO 'test.json'")
duckdb.query("CREATE OR REPLACE TEMP TABLE jobs AS FROM 'test.json'")
duckdb.query("SELECT j.* FROM jobs")

Ideally I would prefer to just do select j.* from job_details which will automatically convert from json to struct.

**Comment by loicalleyne** [comment]
Is there a way to make a formal feature request? Should I create an issue instead of a discussion?

**Comment by themichaelusa** [comment]
Bumping this again!

**Comment by aszenz** [comment]
Just ran into this, we store dynamic data inside a json column, it would be awesome to explode all the keys from it just like read_json_auto does.
I actually think both `read_json_auto` and `read_csv_auto` should work on files and expressions which generate csv/json (one or multiple)

**Comment by FrancoisLepoutre** [comment]
I will concur with **ALL** of you.  Storing json in duckdb tables whether stored within the newly stabilised database format or in ad hoc parquets would be great.

Currently using a couple of tricks in order to avoid having files - everything staying at in-memory duckdb tables or dynamic views - in order to get a fast and secure fully json-controlled application running.

**This feature request is a great addition for those building apps on top of our great duck!**

**Comment by rvs** [comment]
Any updates on this one? It would be really, really helpful for a lot of LLM training usecases I'm dealing with right now

**Comment by loicalleyne** [comment]
ClickHouse recently released support for a _variable_ type that has enabled  a new _json_ type.

**Comment by rahul-sheelavantar** [comment]
Hi, I want to know if I can parse multiple json columns using duckdb

for example
i have 

```
[{
  location: {
    city: STRING,
    state: STRING
  },
  status: {
    isActive: BOOLEAN
  }
}]
```

parse to 

```
{
  location_city: STRING,
  location_state: STRING,
  status_isActive: BOOLEAN
}
```

Or anything close to this is also fine

**Reply by cmdlineluser** [reply]
Looks similar to the unnest with prefix request:
- https://github.com/duckdb/duckdb/discussions/11012
